content/4_Gold/String_Hashing.mdx

```diff
@@ -186,11 +186,150 @@ It's generally a good idea to use two randomized bases rather than just one to d
 
 ## Solution - Searching For Strings
 
-<IncompleteSection>
+**Time Complexity:** $\mathcal O(|N| + |H|)$.
+
+We'll use a [sliding window](/gold/sliding) over $H$ to find the "matches" with $N$.
+
+Since we don't care about relative order when comparing two substrings, we can store frequency tables of the characters in the current window and in $N$. When we slide the window, at most two values in that table change. To compare two substrings, we simply compare the 26 values in each table.
+
+If we only needed to count the number of matches, then the above alone would suffice (in fact, [IOI 2006 Writing](https://dmoj.ca/problem/ioi06p1) is just that). However, we need to count the **distinct** permutations of $N$ in $H$, so we need to be a bit more clever.
+
+One way to solve this is by storing the polynomial hashes of each match in a hashset, since we expect different permutations to have different polynomial hashes. The answer would simply be the size of that hashset at the end.
+
+Since the test data for this particular problem is very strong, we will probably get hash collisions with only one hash. To remedy this, we use two hashes for each match - this significantly decreases the probability of collisions.
+
+The two base-modulo pairs $(9973, 10^9 + 9)$ and $(69420, 998244353)$ work for this problem.
```

Comment:
using the same modulo twice with different bases, or the same base twice with different modulos suffices